#### 1 - Cite 5 diferenças entre o Random Forest e o AdaBoost

#### 1. Método de Combinação dos Modelos
- **Random Forest**:
  - Combina múltiplas árvores de decisão, onde cada árvore é construída de forma independente usando uma amostra aleatória dos dados com reposição (bootstrap sampling).
  - As previsões finais são obtidas pela média (em regressão) ou pela maioria dos votos (em classificação).

- **AdaBoost**:
  - Combina múltiplos modelos fracos (frequentemente árvores de decisão com profundidade 1, chamadas de "stumps"), onde cada modelo subsequente é treinado com um peso maior nos exemplos que foram mal classificados pelos modelos anteriores.
  - As previsões finais são uma combinação ponderada das previsões de todos os modelos.

#### 2. Importância das Amostras
- **Random Forest**:
  - Todas as amostras têm a mesma importância ao construir cada árvore, pois são selecionadas aleatoriamente com reposição.

- **AdaBoost**:
  - A importância das amostras muda a cada iteração. Exemplos mal classificados recebem pesos maiores para que o modelo subsequente preste mais atenção a eles.

#### 3. Robustez ao Overfitting
- **Random Forest**:
  - Geralmente é menos propenso ao overfitting devido à combinação de várias árvores independentes e ao uso de amostragem bootstrap e seleção aleatória de features.

- **AdaBoost**:
  - Pode ser mais propenso ao overfitting, especialmente se o número de iterações (número de modelos fracos) for muito alto e o modelo base for muito complexo.

#### 4. Parâmetros de Controle
- **Random Forest**:
  - Os principais parâmetros incluem o número de árvores (`n_estimators`), o número de features a serem consideradas para dividir em cada nó (`max_features`), e a profundidade máxima das árvores (`max_depth`).

- **AdaBoost**:
  - Os principais parâmetros incluem o número de iterações (`n_estimators`) e a taxa de aprendizado (`learning_rate`), que controla a contribuição de cada modelo fraco.

#### 5. Interpretação das Importâncias das Features
- **Random Forest**:
  - Fornece uma medida direta da importância das features, baseada na redução da impureza (por exemplo, Gini ou entropia) causada por cada feature em todas as árvores.

- **AdaBoost**:
  - A importância das features pode ser interpretada com base nos pesos atribuídos aos modelos fracos e na frequência com que uma feature é escolhida para divisão, mas é menos direta e mais complicada do que no Random Forest.


### Abaixo, uma explicação didática quanto ao funcionamento do Adaboost:

#### Funcionamento dos Pesos no AdaBoost

#### Inicialização dos Pesos dos Exemplos
Imagine que temos um conjunto de dados com 10 exemplos. Inicialmente, cada exemplo tem um peso igual de 0.1 (já que \( \frac{1}{10} = 0.1 \)).

#### Treinamento do Primeiro Modelo Fraco
- Treinamos o primeiro modelo fraco (uma árvore de decisão simples) usando os dados ponderados.
- O modelo faz previsões. Alguns exemplos serão classificados corretamente e outros incorretamente.

#### Cálculo do Erro do Modelo e do Peso do Modelo
- Calculamos a taxa de erro do modelo, que é a soma dos pesos dos exemplos que foram classificados incorretamente.
- Usamos essa taxa de erro para calcular o peso do modelo. Modelos com menor erro recebem um peso maior.

#### Atualização dos Pesos dos Exemplos
- Ajustamos os pesos dos exemplos:
  - Exemplos classificados incorretamente pelo modelo têm seus pesos aumentados.
  - Exemplos classificados corretamente têm seus pesos diminuídos.
- Isso significa que exemplos difíceis, que foram mal classificados, terão mais importância na próxima iteração.

#### Repetição do Processo
- Treinamos o próximo modelo fraco com os dados ponderados ajustados.
- Repetimos o processo de ajuste de pesos para os exemplos e os modelos.


### 2. Acesse o link Scikit-learn– adaboost , leia a explicação (traduza se for preciso) e crie um jupyter notebook contendo o exemplo do AdaBoost.

In [2]:
from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_iris
from sklearn.ensemble import AdaBoostClassifier

# Carrega o conjunto de dados Iris
X, y = load_iris(return_X_y=True)

# Cria um classificador AdaBoost
clf = AdaBoostClassifier(n_estimators=100, algorithm="SAMME")

# Realiza validação cruzada com 5 divisões (folds)
scores = cross_val_score(clf, X, y, cv=5)

# Calcula e imprime a média dos scores de validação cruzada
scores.mean()

0.9533333333333334

**Observações**

## Hiperparâmetros do AdaBoostClassifier

No exemplo fornecido, utilizamos o `AdaBoostClassifier` com os seguintes hiperparâmetros:

- **base_estimator**: Especifica o estimador base a ser usado. No exemplo, utilizamos uma árvore de decisão rasa (`DecisionTreeClassifier(max_depth=1)`). As árvores de decisão rasas são fracas por si só, mas ao serem combinadas, podem formar um classificador forte.
- **n_estimators**: Define o número de estimadores fracos a serem combinados. Utilizamos 100 estimadores. Este valor é um compromisso comum que proporciona um bom equilíbrio entre bias e variância, melhorando a performance sem exigir recursos computacionais excessivos.
- **algorithm**: Especifica o algoritmo a ser usado. Utilizamos "SAMME", que é adequado para classificação multiclasse e utiliza previsões de classes discretas.

## Validação Cruzada

Utilizamos a validação cruzada para avaliar o desempenho do modelo. No exemplo, aplicamos validação cruzada com 5 divisões (folds), onde o conjunto de dados é dividido em 5 partes. O modelo é treinado e testado 5 vezes, cada vez utilizando uma parte diferente como conjunto de teste e as outras partes como conjunto de treinamento. 

```python
scores = cross_val_score(clf, X, y, cv=5)

**Interpretação do Resultado:**

Em média, o classificador AdaBoost com os parâmetros especificados atingiu uma precisão de aproximadamente 95.33% nos diferentes subconjuntos de teste durante a validação cruzada. Este resultado indica que o modelo é bastante preciso na classificação do conjunto de dados Iris, sendo capaz de identificar corretamente a maioria das amostras.

#### 3. Cite 5 Hyperparametros importantes no AdaBoost.

# Hiperparâmetros Importantes no AdaBoost

Os cinco hiperparâmetros mais importantes no AdaBoost são `n_estimators`, `learning_rate`, `base_estimator`, `algorithm`, e `random_state`.

1. **`n_estimators`**: Este hiperparâmetro define o número de estimadores fracos (por exemplo, árvores de decisão rasas) a serem combinados. Um número maior de estimadores pode melhorar a performance do modelo ao reduzir o bias, mas também pode aumentar o risco de overfitting e o tempo de treinamento. Um valor comum de partida é 100. Exemplo: `AdaBoostClassifier(n_estimators=100)`.

2. **`learning_rate`**: A taxa de aprendizado reduz a contribuição de cada modelo fraco, controlando a influência de cada estimador na predição final. Uma taxa de aprendizado menor pode exigir mais estimadores para alcançar uma boa performance, enquanto uma taxa maior pode acelerar a convergência, mas aumenta o risco de overfitting. Exemplo: `AdaBoostClassifier(learning_rate=1.0)`.

3. **`base_estimator`**: Este hiperparâmetro especifica o modelo de base (fraco) a ser utilizado. Frequentemente é uma árvore de decisão rasa (`DecisionTreeClassifier(max_depth=1)`). A escolha do estimador base pode impactar significativamente a performance e a complexidade do modelo. Exemplo: `AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=1))`.

4. **`algorithm`**: Define o algoritmo de boosting a ser utilizado. Pode ser "SAMME" ou "SAMME.R". "SAMME" utiliza previsões de classes discretas, enquanto "SAMME.R" utiliza previsões probabilísticas, tendendo a ter melhor desempenho em muitos casos. Exemplo: `AdaBoostClassifier(algorithm="SAMME.R")`.

5. **`random_state`**: Este hiperparâmetro controla a semente do gerador de números aleatórios, garantindo a reprodutibilidade dos resultados. Definir o `random_state` permite que os resultados sejam replicados em execuções subsequentes. Exemplo: `AdaBoostClassifier(random_state=42)`.

A escolha adequada e a otimização desses parâmetros através de técnicas como validação cruzada e Grid Search podem melhorar significativamente os resultados do modelo.


#### 4 -(Opcional) Utilize o GridSearch para encontrar os melhores hyperparametros para o conjunto de dados do exemplo (load_iris)

In [4]:
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.datasets import load_iris
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

# Carrega o conjunto de dados Iris
X, y = load_iris(return_X_y=True)

# Cria um classificador base (árvore de decisão rasa)
estimator = DecisionTreeClassifier(max_depth=1)

# Cria um classificador AdaBoost
clf = AdaBoostClassifier(estimator=estimator, algorithm="SAMME")

# Define a grade de hiperparâmetros para busca
param_grid = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 1.0],
    'algorithm': ['SAMME', 'SAMME.R']
}

# Cria o GridSearchCV com validação cruzada
grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy')

# Ajusta o modelo aos dados
grid_search.fit(X, y)

# Imprime os melhores hiperparâmetros encontrados
print(f"Melhores parâmetros: {grid_search.best_params_}")
print(f"Melhor score: {grid_search.best_score_}")

# Avaliação do modelo com os melhores hiperparâmetros
best_clf = grid_search.best_estimator_
scores = cross_val_score(best_clf, X, y, cv=5)
print(f"Média dos scores de validação cruzada: {scores.mean()}")


c:\Users\giova\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
c:\Users\giova\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
c:\Users\giova\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(
c:\Users\giova\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\ensemble\_weight_boosting.py:527: FutureWarning: The SAMME.R algorithm (the default) is 

Melhores parâmetros: {'algorithm': 'SAMME', 'learning_rate': 0.1, 'n_estimators': 200}
Melhor score: 0.96
Média dos scores de validação cruzada: 0.96


# Otimização de Hiperparâmetros com GridSearchCV

Neste exemplo, utilizamos o GridSearchCV para otimizar os hiperparâmetros do `AdaBoostClassifier` no conjunto de dados Iris. O objetivo é melhorar a performance do modelo ajustando os parâmetros para obter a melhor acurácia possível.

## Descrição do Código

1. **Carregamento do Conjunto de Dados**
   - Utilizamos o conjunto de dados Iris, disponível no scikit-learn.
   - Este conjunto de dados é carregado e dividido em características (X) e rótulos (y).

2. **Definição do Classificador Base**
   - Utilizamos uma árvore de decisão rasa (`DecisionTreeClassifier` com `max_depth=1`) como estimador base.
   - Esta escolha é comum em AdaBoost, pois árvores rasas são modelos fracos que, quando combinados, podem formar um classificador forte.

3. **Criação do Classificador AdaBoost**
   - Criamos um classificador `AdaBoostClassifier` utilizando o estimador base definido e o algoritmo "SAMME".

4. **Definição da Grade de Hiperparâmetros**
   - Definimos uma grade de hiperparâmetros para buscar os melhores valores:
     - `n_estimators`: [50, 100, 200]
     - `learning_rate`: [0.01, 0.1, 1.0]
     - `algorithm`: ['SAMME', 'SAMME.R']
   - Esses parâmetros controlam o número de estimadores, a taxa de aprendizado e o algoritmo de boosting utilizado.

5. **Busca de Hiperparâmetros com GridSearchCV**
   - Utilizamos o `GridSearchCV` com validação cruzada de 5 folds para encontrar a melhor combinação de hiperparâmetros.
   - Ajustamos o modelo aos dados e avaliamos a performance.

6. **Impressão dos Resultados**
   - Imprimimos os melhores hiperparâmetros encontrados e o melhor score obtido.
   - Avaliamos o modelo com os melhores hiperparâmetros usando validação cruzada e calculamos a média dos scores.

## Resultados

Os melhores hiperparâmetros encontrados foram:
- `algorithm`: 'SAMME'
- `learning_rate`: 0.1
- `n_estimators`: 200

O melhor score foi 0.96, e a média dos scores de validação cruzada foi 0.96.

## Comparação com Resultados Anteriores

No código anterior, utilizamos os seguintes parâmetros:
- `n_estimators=100`
- `algorithm="SAMME"`

A média dos scores de validação cruzada foi 0.9533333333333334.

### Análise Comparativa

A otimização dos hiperparâmetros com o GridSearchCV resultou em uma melhora na performance do modelo:
- A média dos scores de validação cruzada aumentou de 0.9533333333333334 para 0.96.
- Este aumento indica que a combinação otimizada de hiperparâmetros (200 estimadores, taxa de aprendizado de 0.1 e algoritmo "SAMME") proporciona uma classificação mais precisa no conjunto de dados Iris.

### Conclusão

A utilização do GridSearchCV para otimizar os hiperparâmetros do AdaBoostClassifier demonstrou uma melhora significativa na performance do modelo. Ajustar parâmetros como `n_estimators`, `learning_rate` e `algorithm` é crucial para maximizar a acurácia do modelo e evitar problemas de overfitting ou underfitting. A abordagem de validação cruzada garante que os resultados sejam robustos e generalizáveis para novos dados.
